File: ApogeeVGC/sim/dex-abilities/AbilityFlags.cpp
#include "AbilityFlags.h"

AbilityFlags::AbilityFlags(const AbilityFlags& other)
{
	if (other.breakable)
		breakable = std::make_unique<bool>(*other.breakable);

	if (other.cantsuppress)
		cantsuppress = std::make_unique<bool>(*other.cantsuppress);

	if (other.failroleplay)
		failroleplay = std::make_unique<bool>(*other.failroleplay);

	if (other.failskillswap)
		failskillswap = std::make_unique<bool>(*other.failskillswap);

	if (other.noentrain)
		noentrain = std::make_unique<bool>(*other.noentrain);

	if (other.noreceiver)
		noreceiver = std::make_unique<bool>(*other.noreceiver);

	if (other.notrace)
		notrace = std::make_unique<bool>(*other.notrace);

	if (other.notransform)
		notransform = std::make_unique<bool>(*other.notransform);
}

File: ApogeeVGC/Sim/GameObjects/AbilityFlags.cs
namespace ApogeeVGC.Sim.GameObjects;

public record AbilityFlags
{
    // Can be suppressed by Mold Breaker and related effects
    public bool? Breakable { get; init; }
    // Ability can't be suppressed by e.g. Gastro Acid or Neutralizing Gas
    public bool? CantSuppress { get; init; }
    // Role Play fails if target has this Ability
    public bool? FailRolePlay { get; init; }
    // Skill Swap fails if either the user or target has this Ability
    public bool? FailSkillSwap { get; init; }
    // Entrainment fails if user has this Ability
    public bool? NoEntrain { get; init; }
    // Receiver and Power of Alchemy will not activate if an ally faints with this Ability
    public bool? NoReceiver { get; init; }
    // Trace cannot copy this Ability
    public bool? NoTrace { get; init; }
    // Disables the Ability if the user is Transformed
    public bool? NoTransform { get; init; }
}